src/main/scala/beam/agentsim/infrastructure/ZonalParkingManager.scala

```diff
@@ -250,6 +241,62 @@ class ZonalParkingManager(
       )
   }
 
+
+  private[this] def buildMapForFilter(resources: mutable.Map[StallAttributes, StallValues]): Map[IndexForFilter, Map[StallAttributes, StallValues]] = {
+    resources.groupBy { case (key, _) =>
+      IndexForFilter(tazId = key.tazId, reservedFor = key.reservedFor)
+    }.map { case (key, value) => key -> value.toMap}
+  }
+
+  private[this] def buildMapForFind(resources: mutable.Map[StallAttributes, StallValues]): Map[IndexForFind, Array[StallValues]] = {
+    resources.groupBy { case (key, _) =>
+      IndexForFind(tazId = key.tazId, parkingType = key.parkingType, pricingModel = key.pricingModel, reservedFor = key.reservedFor)
+    }.map { case (key, map) => key -> map.values.toArray }
+  }
+
+  private def findPooledResources(tazId: Id[TAZ], preferredType: ParkingType,
+                                  reservedFor: ReservedParkingType): Option[(IndexForFind, StallValues)] = {
+    def find(key: IndexForFind): Option[(IndexForFind, StallValues)] = {
+      mapForFind.get(key).flatMap { arr =>
+        arr
+          .find(stallValue => stallValue.numStalls > 0 && stallValue.feeInCents == 0)
+          .map { found => (key, found)
+          }
+      }
+    }
+    def predicate(tazId: Id[TAZ], preferredType: ParkingType, pricingModel: PricingModel,
+                  reservedFor: ReservedParkingType): Boolean = {
+      val key = IndexForFind(tazId = tazId, parkingType = preferredType, pricingModel = pricingModel, reservedFor = reservedFor)
+      find(key).isDefined
+    }
+    // We still have to do two look-ups by key because we don't know exact pricing model
+    allPricingModels.collectFirst { case pricingModel if predicate(tazId, preferredType, pricingModel, reservedFor) =>
```

Comment:
You'll save yourself some allocations by moving the predicate into the logic. Then you can build one object, check if it's defined and get it. However you have to use something more like:

```
allPricingModels
  .view
  .map(pricingModel => find(IndexForFind(tazId, preferredType, pricingModel, reservedFor))
  .filter(_.isDefined)
  .head.get
```
And, you could do this for the other predicate also.

Reply:
I thought about this approach, but decided to go with `collectFirst`. Yes, for `collectFirst` I have to do one extra call. But if what you proposed is more readable(and it should be definitely faster), I can make changes